firedrake/adjoint/assembly.py

```diff
@@ -19,10 +20,19 @@ def wrapper(*args, **kwargs):
             if not annotate:
                 return output
 
+            tape = get_working_tape()
+
+            coeff_form = form.coefficients()
+            for coeff in coeff_form:
```

Comment:
You do not use coeff_form later, so maybe merge line 25 and 26 in one?

Reply:
Same for block_variable in 31

Reply:
Done